File: api/src/main/java/eu/suro/redis/Redis.java
package eu.suro.redis;

import org.jetbrains.annotations.Nullable;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.Pipeline;
import redis.clients.jedis.Transaction;
import redis.clients.jedis.exceptions.JedisException;

import java.sql.Timestamp;
import java.util.UUID;
import java.util.function.Consumer;

/*
    Используется для транзакций с редисом, чтобы запрос точно дошел до редиса
 */
public class Redis {
    private static RedisManager redisManager;

    public static void init(RedisManager redisManager) throws JedisException {
        Redis.redisManager = redisManager;
        sync(stream -> stream.getJedis().info());
    }

    public static RedisManager getManager() {
        return redisManager;
    }

    @Deprecated
    public static void async(RedisConsumer create) {
        getManager().async(create);
    }

    public static void sync(RedisConsumer create) {
        try (Jedis jedis = getManager().getJedisPool().getResource()) {
            create.run(new RedisStream(jedis));
        } catch (JedisException e) {
            e.printStackTrace();
        }
    }

    public static void pipeline(Consumer<Pipeline> pipelineConsumer) {
        try (Jedis jedis = getManager().getJedisPool().getResource()) {
            try (Pipeline pipelined = jedis.pipelined()) {
                pipelineConsumer.accept(pipelined);
            }
        } catch (JedisException e) {
            e.printStackTrace();
        }
    }

    public static void transaction(Consumer<Transaction> transactionConsumer) {
        try (Jedis jedis = getManager().getJedisPool().getResource()) {
            try (Transaction transaction = jedis.multi()) {
                transactionConsumer.accept(transaction);
            }
        } catch (JedisException e) {
            e.printStackTrace();
        }
    }

    @Nullable
    public static <T> T parseRedisData(String data, Class<T> clazz) {
        if (data != null && !data.isEmpty())
            try {
                Object object = data;
                if (clazz == Integer.class)
                    object = Integer.parseInt(data);
                if (clazz == Long.class)
                    object = Long.parseLong(data);
                if (clazz == Double.class)
                    object = Double.parseDouble(data);
                if (clazz == Float.class)
                    object = Float.parseFloat(data);
                if (clazz == Boolean.class)
                    object = data.equals("true");
                if (clazz == Timestamp.class)
                    object = new Timestamp(Long.parseLong(data));
                if (clazz == UUID.class)
                    object = UUID.fromString(data);
                return (T)object;
            } catch (Exception ex) {
                return null;
            }
        return null;
    }

}

File: api/src/main/kotlin/eu/suro/redis/Redis.kt
package eu.suro.redis

import java.sql.Timestamp
import java.util.*
import java.util.function.Consumer

/*
   Используется для транзакций с редисом, чтобы запрос точно дошел до редиса
*/
object Redis {
    lateinit var manager: RedisManager


    fun init(redisManager: RedisManager) {
        manager = redisManager
//        sync { stream: RedisStream -> stream.jedis.info() }
    }

//    /**
//     * Выполнить транзакцию с редисом async
//     * @param create - Создание транзакции
//     */
//    @Deprecated("")
//    fun async(create: RedisConsumer?) {
//        manager!!.async(create)
//    }

//    /**
//     * Выполнить транзакцию с редисом sync
//     * @param create - Создание транзакции
//     */
//    fun sync(create: RedisConsumer) {
//        try {
//            manager!!.jedisPool.getResource().use { jedis -> create.run(RedisStream(jedis)) }
//        } catch (e: JedisException) {
//            e.printStackTrace()
//        }
//    }
//
//    /**
//     * Выполнить pipeline с редисом sync
//     * @param pipelineConsumer - Создание транзакции
//     */
//    fun pipeline(pipelineConsumer: Consumer<Pipeline?>) {
//        try {
//            manager!!.jedisPool.getResource().use { jedis ->
//                jedis.pipelined().use { pipelined ->
//                    pipelineConsumer.accept(
//                        pipelined
//                    )
//                }
//            }
//        } catch (e: JedisException) {
//            e.printStackTrace()
//        }
//    }
//
//    /**
//     * Выполнить транзакцию с редисом sync
//     * @param transactionConsumer - Создание транзакции
//     */
//    fun transaction(transactionConsumer: Consumer<Transaction?>) {
//        try {
//            manager!!.jedisPool.getResource().use { jedis ->
//                jedis.multi().use { transaction ->
//                    transactionConsumer.accept(
//                        transaction
//                    )
//                }
//            }
//        } catch (e: JedisException) {
//            e.printStackTrace()
//        }
//    }
//
//    /**
//     * Парсинг данных с редиса
//     * @param data - Данные
//     * @param clazz - Класс
//     * @return - Объект
//     */
//    fun <T> parseRedisData(data: String?, clazz: Class<T>): T? {
//        return if (data != null && !data.isEmpty()) try {
//            var `object`: Any = data
//            if (clazz == Int::class.java) `object` = data.toInt()
//            if (clazz == Long::class.java) `object` = data.toLong()
//            if (clazz == Double::class.java) `object` = data.toDouble()
//            if (clazz == Float::class.java) `object` = data.toFloat()
//            if (clazz == Boolean::class.java) `object` = data == "true"
//            if (clazz == Timestamp::class.java) `object` = Timestamp(data.toLong())
//            if (clazz == UUID::class.java) `object` = UUID.fromString(data)
//            `object` as T
//        } catch (ex: Exception) {
//            null
//        } else null
//    }
}
